.github/workflows/release.yml

```diff
@@ -1,63 +1,112 @@
 name: Build and upload to PyPI
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.number }}-${{ github.event.ref }}
+  cancel-in-progress: true
+
 # Only build on published releases
 on:
   release:
     types:
       - published
 
 jobs:
-  build:
-    name: Build sdist
-    runs-on: ubuntu-18.04
+  generate-wheels-matrix:
+    name: Generate wheels matrix
+    runs-on: ubuntu-latest
+    outputs:
+      include: ${{ steps.set-matrix.outputs.include }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Install cibuildwheel
+        run: pipx install cibuildwheel==2.13.1
+      - id: set-matrix
+        run: |
+          MATRIX=$(
+            {
+              cibuildwheel --print-build-identifiers --platform linux \
+              | jq -nRc '{"only": inputs, "os": "ubuntu-latest"}' \
+              && cibuildwheel --print-build-identifiers --platform macos \
+              | jq -nRc '{"only": inputs, "os": "macos-latest"}' \
+              && cibuildwheel --print-build-identifiers --platform windows \
+              | jq -nRc '{"only": inputs, "os": "windows-2019"}'
+            } | jq -sc
+          )
+          echo "include=$MATRIX" >> $GITHUB_OUTPUT
+        env:
+          CIBW_BUILD: "cp39-* cp310-* cp311-*"
+          # Skip 32 bit windows builds and musllinux due to lack of numpy wheels
+          CIBW_SKIP: "*-win32 *-musllinux*"
+          CIBW_ARCHS_MACOS: x86_64 arm64
+  build_wheels:
+    name: Build ${{ matrix.os }} ${{ matrix.only }}
+    needs: generate-wheels-matrix
+    strategy:
+      matrix:
+        include: ${{ fromJson(needs.generate-wheels-matrix.outputs.include) }}
+    runs-on: ${{ matrix.os }}
     defaults:
       run:
-        shell: bash -l {0}
-
+        shell: bash
     steps:
       - uses: actions/checkout@v3
-      - uses: conda-incubator/setup-miniconda@v2
         with:
-          activate-environment: test-environment
-          python-version: ${{ matrix.python-version }}
-          channels: conda-forge/label/testing,conda-forge
+          # We need the full history to generate the proper version number
+          fetch-depth: 0
 
-      - name: Install dependencies
-        run: |
-          PACKAGES="cython fiona matplotlib-base numpy pyproj pykdtree scipy"
-          PACKAGES="$PACKAGES owslib pep8 pillow pyshp pytest"
-          PACKAGES="$PACKAGES pytest-xdist setuptools_scm shapely"
-          conda install $PACKAGES
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.10'
+
+      - uses: pypa/cibuildwheel@v2.13.1
+        with:
+          only: ${{ matrix.only }}
+
+      - uses: actions/upload-artifact@v3
+        with:
+          path: ./wheelhouse/*.whl
+
+  build_sdist:
+    name: Build source distribution
```

Comment:
We may, as in Matplotlib's current build, want to build sdist first and instruct `cibuildwheel` to build out of that.

Reply:
I'm a bit out of my element here. I tried to do that with a new commit on a separate branch.
https://github.com/greglucas/cartopy/tree/setup-updates-tmp

Failed job: https://github.com/greglucas/cartopy/actions/runs/5601696424/jobs/10245962338

I think it is failing because the sdist isn't being built quite right. It is not cythonizing to convert from `.pyx` to `.cpp`, where we need to include the `.cpp` in the sdist to distribute into the wheel. I'm not sure where we are supposed to build the extension then, do we just call `python setup.py build_ext` as a separate step to make sure the `.cpp` is created?

Reply:
If it is easier for you to take over some of this, feel free to branch off yourself or push directly to my branch. I'm definitely happy to have some help on this.

Reply:
OK, I think we just needed to call `cythonize()` if we are going down the non-sdist route, so I've reworked the logic a bit in `setup.py` to try and simplify things a bit more for whether we are in the Cython route or the pre-made cpp route.
Here is a CI run from a separate branch on my fork that demonstrates it working now when building the wheels from the sdist: https://github.com/greglucas/cartopy/actions/runs/5612238838/jobs/10269964395

Reply:
Maybe this is out of scope for this PR, but any thoughts about using numba to replace the .pyx file? 
Upside: No need to compile the .pyx file during build and deal with correctly packaging it. Plus numba has support for automatic parallelization. 
Downside: Another dependency.

Reply:
Definitely out of scope for this PR. If you can show that it is faster than Cython or makes things easier in some way, then a PR would definitely be considered!

Reply:
As much as I think Numba is cool, it has its own challenges due to its reliance on bridging LLVM and Python. For instance, Numba didn't support Python 3.11 until May 2023, which IMO is unacceptable for something we'd consider as a dependency. I understand the dev team is aware how problematic this is and intends to do better for future releases, but I'd want to see a good track record going forward before we begin to consider that.

Reply:
Now that we are preparing to build wheels, I wonder if all the work to make Cython optional is all that useful? It was a convenience to not require it, I suppose, but with stuff like build-requires and isolated build venvs *and* the fact that we're going to make wheels, reducing the need to build from source, maybe that isn't so great as it once was.

Reply:
Possibly we could take it out... The Cython docs do mention to use an environment variable `USE_CYTHON` to control this, so I think we are doing similar to their recommendations. Doesn't hurt to keep it I guess?
https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#distributing-cython-modules